docs/proposals/2401-llm-trainer-v2/README.md

```diff
@@ -0,0 +1,463 @@
+            dtype="bf16",
+        ),
+        num_nodes=5,
+    ),
+    runtime_ref=llm_runtime,
+)
+```
+
+## Design Details
+
+### `torchtune` Plugin
+
+As is shown in the [torchtune official document](https://pytorch.org/torchtune/main/tune_cli.html#run-a-recipe) and [source code](https://github.com/pytorch/torchtune/blob/75965d4281b9b76c454630d015221b9933c77bf3/torchtune/_cli/run.py#L113-L118), the distributed training arguments like `--nnodes` and `--nproc_per_node` should be passed ahead of the recipe argument in the command line, and **cannot be passed by the environment variables** in the `PET_XXX` convention. And also, `torchtune` is extremely different from the fine-tuning paradigm of `torchrun` because it is **recipe and config-based**, which may need more mutation operations in the config file. Here is an [example](https://github.com/Electronic-Waste/kubeflow-llm-trainer/blob/main/torchtune-llm-finetuning.yaml).
+
+Thus, we need to implement a new plugin for `torchtune` if we decide to adopt `torchtune` as a launcher for LLM fine-tuning on Kubernetes. And the new plugin should have these abilities:
+
+1. Parse distributed training arguments in TrainJob and TrainingRuntime API, and integrate them with the `tune run` command.
+2. Handle overrides in the `torchtune` fine-tuning configuration file.
+3. Validate some requirements.
+
+### Fine-Tuning Config
+
+We will add the fine-tuning configurations in the `fine_tuning_config` field in `Trainer` dataclass.
+
+| Parameters | What is it? |
+| - | - |
+| dtype | The underlying data type used to represent the model and optimizer parameters. Currently, we support `bf16` and `fp32`. |
+| peft_config | Configuration for the PEFT(Parameter-Efficient Fine-Tuning), including Lora, AdapterPrompt, PrefixTuning, etc. |
+
+```python
+# FineTuningConfig DataClass
+@dataclass
+class FineTuningConfig:
+    dtype: string = "bf16"
+    peft_config: Optional[Union[LoraConfig]] = None
+
+```
+
+The Python SDK will look like:
+
+**LoRA Config**
+
+The *LoraConfig* represents the config of LoRA we use to fine-tune the model.
+
+| Parameters | What is it? |
+| - | - |
+| apply_lora_to_mlp | Whether to apply LoRA to the MLP in each transformer layer |
+| apply_lora_to_output | Whether to apply LoRA to the model’s final output projection |
+| lora_attn_modules | A list of strings specifying which layers of the model to apply LoRA: 1. `q_proj` applies LoRA to the query projection layer. 2. `k_proj` applies LoRA to the key projection layer. 3. `v_proj` applies LoRA to the value projection layer. 4. `output_proj` applies LoRA to the attention output projection layer. |
+| lora_rank | The rank of the low rank decomposition. |
+| lora_alpha | The scaling factor that adjusts the magnitude of the low-rank matrices’ output |
+| lora_dropout | The probability of applying Dropout to the low rank updates |
+| quantize_base | Whether to enable model quantization |
+| use_dora | Whether to enable DoRA |
+
+
```

Comment:
It would be nice if you could also add the section of how we are going to structure these recipes.
And how we are going to select the default resources, number of PyTorch nodes.